content/blog/upkeep-testthat-3/index.Rmd

```diff
@@ -0,0 +1,142 @@
+3. If your output got noisy, quiet things down as needed.
+4. The harder part: think about what it means if things are not "all equal" anymore.
+
+### Activation `r emo::ji("rocket")`
+
+To activate you need to do two things in the DESCRIPTION - or you can let `usethis::use_testthat(3)` do it for you:
+
+- Increase the testthat version to `>= 3.0.0`.
+- Set the `Config/testthat/edition` field to `3`.
+
+### Moving on from deprecations `r emo::ji("sparkles")`
+
+The article on testthat 3e contains a [list of deprecated functions](https://testthat.r-lib.org/articles/third-edition.html#deprecations) together with their successors. You can work your way through it, searching for the deprecated function and then replacing it with the most suitable alternative. The first one that list is `context()` as testthat will use the file name instead, ensuring that context and file name are in sync. As such, `context()` does not have a replacement and my first commit after activating the third edition was to remove all calls to `context()`, followed by replacing other deprecated functions and arguments.
+
+![A list of commits starting with "require testthat 3e, followed by removing `context()` and other deprecated functions"](commits.png)
+
+### Warnings and messages `r emo::ji("shushing_face")`
+
+Edition 3 handles warnings and messages differently than edition 2: `expect_warning()` captures at most one warning so if your code generates multiple warnings, they will bubble up now. Messages were previously silently ignored, now they also bubble up. That means the output may be a lot noisier after switching to edition 3. If the warnings or messages are important, you should explicitly capture them. Otherwise you can suppress them to clean up the output and make it easier to focus on what's important. Again, the testthat article has good examples for how to do either.
+
+### Comparing things `r emo::ji("apple")` `r emo::ji("orange")`
+
+The last big change from edition 2 to edition 3 that I want to mention is what is happening under the hood of `expect_equal()` and `expect_identical()`. Edition 3 uses `waldo::compare()` while edition 2 uses `all.equal()`. For the most part that meant changing the argument name from `tol` to `tolerance`, like in my third commit above. 
+
+I did however run into the situation where a test newly failed. Those are the situations where general advise is hard because it depends so much on the context. In my case, I made use of the `ignore_function_env` and `ignore_formula_env` arguments to `waldo::compare()` to exclude those environments from the comparison. Those are probably really neat to know about if you are upgrading a modelling package but not particularly important otherwise. For dials and censored, that solved most of the cases. In one instance, I ended up tweaking the reference value based on theoretical considerations of the model I was dealing with rather than increasing the tolerance. 
```

Comment:
```suggestion
I did, however, run into a situation where a test newly failed. Those are the situations where general advice is hard because it depends so much on the context. In my case, I made use of the `ignore_function_env` and `ignore_formula_env` arguments to [`waldo::compare()`](https://waldo.r-lib.org/reference/compare.html) to exclude those environments from the comparison. Those are probably useful to know about if you are upgrading a modelling package, but not particularly important otherwise. For dials and censored, that solved most of the cases. In one instance, I ended up tweaking the reference value based on theoretical considerations of the model I was dealing with rather than increasing the tolerance. 
```

Reply:
Added the link since you're going into details re. the arguments.